xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,725 @@
+   * (with the port, if any, stripped off). Or sends an RDS request if configured for dynamic
+   * resolution. The response is NACKed if contains invalid data for gRPC's usage. Otherwise, an
+   * ACK request is sent to management server.
+   */
+  private void handleLdsResponse(DiscoveryResponse ldsResponse) {
+    logger.log(Level.FINE, "Received a LDS response: {0}", ldsResponse);
+    checkState(ldsResourceName != null && configWatcher != null,
+        "No LDS request was ever sent. Management server is doing something wrong");
+    adsStream.ldsRespNonce = ldsResponse.getNonce();
+
+    // Unpack Listener messages.
+    Map<String, Listener> listeners = new HashMap<>(ldsResponse.getResourcesCount());
+    try {
+      for (com.google.protobuf.Any res : ldsResponse.getResourcesList()) {
+        Listener l = res.unpack(Listener.class);
+        listeners.put(l.getName(), l);
+      }
+    } catch (InvalidProtocolBufferException e) {
+      adsStream.sendNackRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName),
+          ldsResponse.getNonce(), "Broken LDS response");
+      configWatcher.onError(Status.fromThrowable(e).augmentDescription("Broken LDS response"));
+      return;
+    }
+
+    // Unpack HttpConnectionManager messages
+    HttpConnectionManager requestedHttpConnManager = null;
+    List<HttpConnectionManager> httpConnectionManagers = new ArrayList<>();
+    try {
+      for (Map.Entry<String, Listener> entry : listeners.entrySet()) {
+        HttpConnectionManager hm =
+            entry.getValue().getApiListener().getApiListener().unpack(HttpConnectionManager.class);
+        httpConnectionManagers.add(hm);
+        if (entry.getKey().equals(ldsResourceName)) {
+          requestedHttpConnManager = hm;
+        }
+      }
+    } catch (InvalidProtocolBufferException e) {
+      adsStream.sendNackRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName),
+          ldsResponse.getNonce(), "Broken LDS response");
+      configWatcher.onError(Status.fromThrowable(e).augmentDescription("Broken LDS response"));
+      return;
+    }
+
+    boolean invalidData = false;
+    // All RouteConfigurations referenced by this LDS response, either in in-lined
+    // RouteConfiguration message or in RDS config.
+    Set<String> routeConfigs = new HashSet<>();
+    for (HttpConnectionManager hm : httpConnectionManagers) {
+      // The HttpConnectionManager message must either provide the RouteConfiguration directly
+      // in-line or tell the client to use RDS to obtain it.
+      if (hm.hasRouteConfig()) {
+        routeConfigs.add(hm.getRouteConfig().getName());
+      } else if (hm.hasRds()) {
+        Rds rds = hm.getRds();
+        // For using RDS, it must be set to use ADS.
+        if (!rds.getConfigSource().hasAds()) {
+          invalidData = true;
+          break;
+        }
+        routeConfigs.add(rds.getRouteConfigName());
+      } else {
+        invalidData = true;
+        break;
+      }
+    }
+
```

Comment:
At this stage if `invalidData` is true why do you want to execute code between lines 282 and 298? You are anyway bailing out on line 300.

This is why I wanted to see smaller functions and specialized processor classes.

Reply:
Fixed.